snapcraft/_store.py

```diff
@@ -375,39 +379,104 @@ def sign_build(snap_filename, key_name=None, local=False):
+
+    if os.environ.get('DELTA_UPLOADS_EXPERIMENTAL') and source_snap:
+        try:
+            result = _upload_delta(snap_name, snap_filename, source_snap)
+        except StoreDeltaApplicationError:
+            logger.warning('Unable to upload delta, '
+                           'falling back to uploading full snap.')
+            result = _upload_snap(snap_name, snap_filename)
+    else:
+        result = _upload_snap(snap_name, snap_filename)
+
     # This is workaround until LP: #1599875 is solved
     if 'revision' in result:
-        logger.info('Revision {!r} of {!r} created.'.format(
-            result['revision'], snap_name))
+        logger.info('Revision {!r} of {!r} created'.format(
+            result['revision'], snap_name))\
+
+        if os.environ.get('DELTA_UPLOADS_EXPERIMENTAL'):
+            snap_cache.cache(snap_filename, result['revision'])
+            snap_cache.prune(keep_revision=result['revision'])
     else:
         logger.info('Uploaded {!r}'.format(snap_name))
-    tracker.raise_for_code()
-
-    if os.environ.get('DELTA_UPLOADS_EXPERIMENTAL'):
-        snap_cache = cache.SnapCache(project_name=snap_name)
-        snap_cache.cache(snap_filename, result['revision'])
-        snap_cache.prune(keep_revision=result['revision'])
 
     if release_channels:
         release(snap_name, result['revision'], release_channels)
 
 
+def _upload_snap(snap_name, snap_filename):
+    store = storeapi.StoreClient()
+    logger.info('Uploading {}'.format(snap_filename))
+    with _requires_login():
+        tracker = store.upload(snap_name, snap_filename)
+    result = tracker.track()
+    tracker.raise_for_code()
+    return result
+
+
+def _upload_delta(snap_name, snap_filename, source_snap):
+    store = storeapi.StoreClient()
+    delta_format = 'xdelta3'
+    logger.info('Found cached source snap {}'.format(source_snap))
+    target_snap = os.path.join(os.getcwd(), snap_filename)
+
+    try:
+        xdelta_generator = deltas.XDelta3Generator(
+            source_path=source_snap, target_path=target_snap)
+        delta_filename = xdelta_generator.make_delta()
+    except DeltaGenerationError:
+        raise StoreDeltaApplicationError
+
+    snap_hashes = {'source_hash': calculate_sha3_384(source_snap),
+                   'target_hash': calculate_sha3_384(target_snap),
+                   'delta_hash': calculate_sha3_384(delta_filename)}
+
+    try:
+        logger.info('Uploading delta {}'.format(delta_filename))
+        with _requires_login():
+            delta_tracker = store.upload(
+                snap_name,
+                delta_filename,
+                delta_format=delta_format,
+                source_hash=snap_hashes['source_hash'],
+                target_hash=snap_hashes['target_hash'],
+                delta_hash=snap_hashes['delta_hash'])
+        result = delta_tracker.track()
+        delta_tracker.raise_for_code()
+    except storeapi.errors.StoreReviewError as e:
```

Comment:
we've been putting this logic in `storeapi.errors`, what are the gains of going this path?